File: go/internal/cmd/resolve.go
package cmd

import (
	"fmt"

	"github.com/spf13/cobra"

	"git.zib.de/cspiegel/threads/internal/git"
	"git.zib.de/cspiegel/threads/internal/thread"
	"git.zib.de/cspiegel/threads/internal/workspace"
)

var (
	resolveCommit bool
	resolveMsg    string
)

var resolveCmd = &cobra.Command{
	Use:   "resolve <id>",
	Short: "Mark thread resolved",
	Args:  cobra.ExactArgs(1),
	RunE:  runResolve,
}

func init() {
	resolveCmd.Flags().BoolVar(&resolveCommit, "commit", false, "Commit after resolving")
	resolveCmd.Flags().StringVarP(&resolveMsg, "m", "m", "", "Commit message")
}

func runResolve(cmd *cobra.Command, args []string) error {
	ws := getWorkspace()
	ref := args[0]

	file, err := workspace.FindByRef(ws, ref)
	if err != nil {
		return err
	}

	t, err := thread.Parse(file)
	if err != nil {
		return err
	}

	oldStatus := t.Status()

	// Update status
	if err := t.SetFrontmatterField("status", "resolved"); err != nil {
		return err
	}

	// Add log entry
	t.Content = thread.InsertLogEntry(t.Content, "Resolved.")

	if err := t.Write(); err != nil {
		return err
	}

	fmt.Printf("Resolved: %s → resolved (%s)\n", oldStatus, file)

	if resolveCommit {
		msg := resolveMsg
		if msg == "" {
			msg = git.GenerateCommitMessage(ws, []string{file})
		}
		if err := git.AutoCommit(ws, file, msg); err != nil {
			return err
		}
	} else {
		fmt.Printf("Note: Thread %s has uncommitted changes. Use 'threads commit %s' when ready.\n", ref, ref)
	}

	return nil
}

File: src/cmd/resolve.rs
use std::path::Path;

use clap::Args;
use clap_complete::engine::ArgValueCompleter;
use serde::Serialize;

use crate::args::FormatArgs;
use crate::config::{Config, env_bool, is_quiet};
use crate::git;
use crate::output::{self, OutputFormat};
use crate::thread::Thread;
use crate::workspace;

#[derive(Args)]
pub struct ResolveArgs {
    /// Thread ID or name reference
    #[arg(add = ArgValueCompleter::new(crate::workspace::complete_thread_ids))]
    id: String,

    /// Commit after resolving
    #[arg(long)]
    commit: bool,

    /// Commit message
    #[arg(short = 'm', long)]
    m: Option<String>,

    #[command(flatten)]
    format: FormatArgs,
}

#[derive(Serialize)]
struct ResolveOutput {
    id: String,
    old_status: String,
    path: String,
    committed: bool,
}

pub fn run(args: ResolveArgs, ws: &Path, config: &Config) -> Result<(), String> {
    let format = args.format.resolve();

    let file = workspace::find_by_ref(ws, &args.id)?;

    let mut t = Thread::parse(&file)?;

    let old_status = t.status().to_string();
    let id = t.id().to_string();

    // Update status using config default
    let closed_status = &config.defaults.closed;
    t.set_frontmatter_field("status", closed_status)?;

    // Add log entry
    let log_msg = if closed_status == "resolved" {
        "Resolved.".to_string()
    } else {
        format!("Closed ({}).", closed_status)
    };
    t.insert_log_entry(&log_msg)?;

    t.write()?;

    let should_commit = args.commit || env_bool("THREADS_AUTO_COMMIT").unwrap_or(false);
    let committed = if should_commit {
        let repo = workspace::open()?;
        let rel_path = file.strip_prefix(ws).unwrap_or(&file);
        let msg = args
            .m
            .unwrap_or_else(|| git::generate_commit_message(&repo, &[rel_path]));
        git::auto_commit(&repo, &file, &msg)?;
        true
    } else {
        false
    };

    let rel_path = workspace::path_relative_to_git_root(ws, &file);

    match format {
        OutputFormat::Pretty | OutputFormat::Plain => {
            println!("Closed: {} → {} ({})", old_status, closed_status, rel_path);
            if !committed && !is_quiet(config) {
                output::print_uncommitted_hint(&id);
            }
        }
        OutputFormat::Json => {
            let output = ResolveOutput {
                id,
                old_status,
                path: rel_path,
                committed,
            };
            let json = serde_json::to_string_pretty(&output)
                .map_err(|e| format!("JSON serialization failed: {}", e))?;
            println!("{}", json);
        }
        OutputFormat::Yaml => {
            let output = ResolveOutput {
                id,
                old_status,
                path: rel_path,
                committed,
            };
            let yaml = serde_yaml::to_string(&output)
                .map_err(|e| format!("YAML serialization failed: {}", e))?;
            print!("{}", yaml);
        }
    }

    Ok(())
}
